syntax = "proto3";

package aea.eightballer.spot_asset.v0_1_0;

message SpotAssetMessage{

  // Custom Types
  message Decimal{
    int32 Base = 1;
    float Float = 2;
    string Display = 3;
  }

  message ErrorCode{
    enum ErrorCodeEnum {
      ERROR_CODE_ENUM_UNSUPPORTED_PROTOCOL = 0;
      ERROR_CODE_ENUM_DECODING_ERROR = 1;
      ERROR_CODE_ENUM_INVALID_MESSAGE = 2;
      ERROR_CODE_ENUM_UNSUPPORTED_SKILL = 3;
      ERROR_CODE_ENUM_INVALID_DIALOGUE = 4;
    }
    ErrorCodeEnum error_code = 1;
  }


  // Performatives and contents
  message Get_Spot_Asset_Performative{
    string name = 1;
    string exchange_id = 2;
  }

  message Spot_Asset_Performative{
    string name = 1;
    Decimal total = 2;
    Decimal free = 3;
    Decimal available_without_borrow = 4;
    Decimal usd_value = 5;
    bool usd_value_is_set = 6;
    Decimal decimal = 7;
    bool decimal_is_set = 8;
  }

  message Get_Spot_Assets_Performative{
    string exchange_id = 1;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
  }

  message End_Performative{
  }


  oneof performative{
    End_Performative end = 5;
    Error_Performative error = 6;
    Get_Spot_Asset_Performative get_spot_asset = 7;
    Get_Spot_Assets_Performative get_spot_assets = 8;
    Spot_Asset_Performative spot_asset = 9;
  }
}
